### PR TITLE
Fix endianness problem in sshkdf

### PR DESCRIPTION
**Pull request by juergenchrist**
Handling of parameter OSSL_KDF_PARAM_SSHKDF_TYPE caused endianness problems on
s390.  It is not possible to pass FIPS tests since the parameter was stored
with an integer value but read via a cast to char pointer.  While this works
on little endian machines, big endian s390 read the most significant bits
instead of the least significant (as done by, e.g., x86).  Change the
parameter to char array and fix the usages.

Signed-off-by: Juergen Christ <jchrist@linux.ibm.com>

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
Wouldn't it make actually more sense to fix this at the provider side? To treat the parameter as integer and not as an UTF8 string.

**Comment by juergenchrist** [comment]
The tests seemed to indicate that this should actually be a one-letter string consisting of a letter from the range 'A' to 'F'.  But of course we could also change the tests to provide the correct integer value.  See file test/recipes/30-test_evp_data/evpkdf_ssh.txt (`Ctrl.type = type:A` for example)

**Comment by t8m** [comment]
And the documentation would have to be changed too but in a different way. So it is really a matter of taste which approach would be better.

@openssl/otc Any other opinion? Also shouldn't we actually error out if a numeric parameter is treated as string or vice-versa?

**Comment by t8m** [comment]
Also please to accept this we would need your ICLA submitted.

**Comment by juergenchrist** [comment]
> Also please to accept this we would need your ICLA submitted.

That is already done, but the company CCLA update is still pending...

**Comment by mattcaswell** [comment]
> Any other opinion? 

I have no strong view. Maybe @slontis does...?

**Comment by slontis** [comment]
I am on holidays till next week.
I would just leave this as a integer and address the issue it has with using integers.

**Comment by paulidale** [comment]
Character set concerns appear if it is changed to a UTF-8 string.  EBCDIC vs ASCII primarily.
I agree with @slontis's approach.

**Comment by mattcaswell** [comment]
> That is already done, but the company CCLA update is still pending...

CCLA has been received but it only lists one of your email addresses, and not your "linux.ibm.com" one (just your "ibm.com" one) - so CLA check is still going to fail. Can we either get CCLA changed to list all valid emails, or update the author field in the commits here?

**Comment by slontis** [comment]
If this is really an integer then it should be doing something like the following inside kdf_sshkdf_set_ctx_params().

```
if (!OSSL_PARAM_get_int(p, &t)
    || t < EVP_KDF_SSHKDF_TYPE_INITIAL_IV_CLI_TO_SRV
    || t > EVP_KDF_SSHKDF_TYPE_INTEGRITY_KEY_SRV_TO_CLI) {
```
instead of doing
```
t = *(unsigned char *)p->data;
if (t < 65 || t > 70) {
```

**Comment by levitte** [comment]
Hmmmm, this is originally a size issue, isn't it?  Doesn't the recently merged #13429 help?

**Comment by levitte** [comment]
@t8m sez:
> Also shouldn't we actually error out if a numeric parameter is treated as string or vice-versa?

Yes, but it seems we don't check the data type in that particular case...  no idea why.

https://github.com/openssl/openssl/blob/e211d949cd5737e53cd3399e6a88453930768b98/providers/implementations/kdfs/sshkdf.c#L157-L167

**Comment by levitte** [comment]
> Character set concerns appear if it is changed to a UTF-8 string. EBCDIC vs ASCII primarily.

The owner of the used OSSL_PARAM array needs to ensure that it's actual UTF-8, i.e. convert from EBCDIC to ASCII (since the latter is UTF-8 compatible)

**Comment by levitte** [comment]
Considering A to F, could it be that the original intent was hex?

**Comment by juergenchrist** [comment]
> Considering A to F, could it be that the original intent was hex?

No, it really is ASCII A to F (see https://tools.ietf.org/html/rfc4253#section-7.2)

**Comment by levitte** [comment]
> > Considering A to F, could it be that the original intent was hex?
> 
> No, it really is ASCII A to F (see https://tools.ietf.org/html/rfc4253#section-7.2)

Thank you!  I was trying to find this info, but my google-fu failed me.

They are presented as ASCII strings in the RFC...  in that case, I fail to see why we should make them integers...  for EBCDIC, it's not *that* hard to convert (since we have the tools), and the provider side already does the right thing by picking out the first character and comparing it with ASCII codes, all that's lacking there is a check that the data type is `OSSL_PARAM_UTF8_STRING`.

**Comment by rsbeckerca** [comment]
Really hoping this one gets merged soon.

**Comment by rsbeckerca** [comment]
@juergenchrist Have you made any headway on this?

**Comment by levitte** [comment]
Having looked at this a number of times, and also played around with an integer solution, I still lean toward this PR over that, most of all because of usability concerns.  Someone playing around with SSHKDF will no doubt have read RFC 4253, or other descriptions derived from that RFC, and will no doubt have heard that there are types, denoted with a single ASCII character.

If we change the type according to docs, i.e. make it an actual integer, these are examples of usability consequences:

``` diff
diff --git a/test/recipes/20-test_kdf.t b/test/recipes/20-test_kdf.t
index d0ea07ee9a..8d40b35b42 100755
--- a/test/recipes/20-test_kdf.t
+++ b/test/recipes/20-test_kdf.t
@@ -37,7 +37,7 @@ my @kdf_tests = (
     { cmd => [qw{openssl kdf -keylen 14 -kdfopt digest:SHA224 -kdfopt hexkey:6dbdc23f045488e4062757b06b9ebae183fc5a5946d80db93fec6f62ec07e3727f0126aed12ce4b262f47d48d54287f81d474c7c3b1850e9 -kdfopt hexinfo:a1b2c3d4e54341565369643c832e9849dcdba71e9a3139e606e095de3c264a66e98a165854cd07989b1ee0ec3f8dbe SSKDF}],
       expected => 'a4:62:de:16:a8:9d:e8:46:6e:f5:46:0b:47:b8',
       desc => 'SSKDF HASH SHA224'},
-    { cmd => [qw{openssl kdf -keylen 16 -kdfopt digest:SHA256 -kdfopt hexkey:0102030405 -kdfopt hexxcghash:06090A -kdfopt hexsession_id:01020304 -kdfopt type:A SSHKDF}],
+    { cmd => [qw{openssl kdf -keylen 16 -kdfopt digest:SHA256 -kdfopt hexkey:0102030405 -kdfopt hexxcghash:06090A -kdfopt hexsession_id:01020304 -kdfopt hextype:41 SSHKDF}],
     expected => '5C:49:94:47:3B:B1:53:3A:58:EB:19:42:04:D3:78:16',
     desc => 'SSHKDF SHA256'},
 );
```

``` diff
diff --git a/test/recipes/30-test_evp_data/evpkdf_ssh.txt b/test/recipes/30-test_evp_data/evpkdf_ssh.txt
index d870df7037..8a49d03e6d 100644
--- a/test/recipes/30-test_evp_data/evpkdf_ssh.txt
+++ b/test/recipes/30-test_evp_data/evpkdf_ssh.txt
@@ -19,7 +19,8 @@ Ctrl.digest = digest:SHA1
 Ctrl.hexkey = hexkey:0000008055bae931c07fd824bf10add1902b6fbc7c665347383498a686929ff5a25f8e40cb6645ea814fb1a5e0a11f852f86255641e5ed986e83a78bc8269480eac0b0dfd770cab92e7a28dd87ff452466d6ae867cead63b366b1c286e6c4811a9f14c27aea14c5171d49b78c06e3735d36e6a3be321dd5fc82308f34ee1cb17fba94a59
 Ctrl.hexxcghash = hexxcghash:a4ebd45934f56792b5112dcd75a1075fdc889245
 Ctrl.hexsession_id = hexsession_id:a4ebd45934f56792b5112dcd75a1075fdc889245
-Ctrl.type = type:A
+# 41 = hex ASCII code for "A"
+Ctrl.hextype = hextype:41
 Output = e2f627c0b43f1ac1
 
 KDF = SSHKDF
```

Forcing the users to constantly re-calculate a letter to hex isn't exactly friendly...

I would, however, add this change:

``` diff
diff --git a/providers/implementations/kdfs/sshkdf.c b/providers/implementations/kdfs/sshkdf.c
index daf0dd2e87..e86c502184 100644
--- a/providers/implementations/kdfs/sshkdf.c
+++ b/providers/implementations/kdfs/sshkdf.c
@@ -135,7 +135,6 @@ static int kdf_sshkdf_set_ctx_params(void *vctx, const OSSL_PARAM params[])
     const OSSL_PARAM *p;
     KDF_SSHKDF *ctx = vctx;
     OSSL_LIB_CTX *provctx = PROV_LIBCTX_OF(ctx->provctx);
-    int t;
 
     if (!ossl_prov_digest_load_from_params(&ctx->digest, params, provctx))
         return 0;
@@ -156,14 +155,17 @@ static int kdf_sshkdf_set_ctx_params(void *vctx, const OSSL_PARAM params[])
 
     if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_SSHKDF_TYPE))
         != NULL) {
-        if (p->data == NULL || p->data_size == 0)
+        const char *kdftype;
+
+        if (!OSSL_PARAM_get_utf8_string_ptr(p, &kdftype))
+            return 0;
+        if (kdftype == NULL || kdftype[0] == '\0' || kdftype[1] != '\0')
             return 0;
-        t = *(unsigned char *)p->data;
-        if (t < 65 || t > 70) {
+        if (kdftype[0] < 65 || kdftype[0] > 70) {
             ERR_raise(ERR_LIB_PROV, PROV_R_VALUE_ERROR);
             return 0;
         }
-        ctx->type = (char)t;
+        ctx->type = kdftype[0];
     }
     return 1;
 }
```
Regarding the issue of EBCDIC, it turns out to be easy to solve.  PR coming up!

**Comment by juergenchrist** [comment]
I added the patch from @levitte regarding the additional checks.  Not sure about the documentation changes mentioned by @t8m.  Can anybody help me out here?  Also I cannot reproduce the error in the GitHub CI locally.  For me, 30-test_evp_kdf.t passes without failure.

**Comment by levitte** [comment]
It may be that this diff will fix the failures:

``` diff
diff --git a/test/evp_kdf_test.c b/test/evp_kdf_test.c
index d56e14cdb0..ccb930a881 100644
--- a/test/evp_kdf_test.c
+++ b/test/evp_kdf_test.c
@@ -1207,7 +1207,9 @@ static int test_kdf_sshkdf(void)
     int ret;
     EVP_KDF_CTX *kctx;
     OSSL_PARAM params[6], *p = params;
-    char kdftype = EVP_KDF_SSHKDF_TYPE_INITIAL_IV_CLI_TO_SRV;
+    char kdftype[] = {
+        EVP_KDF_SSHKDF_TYPE_INITIAL_IV_CLI_TO_SRV, 0
+    };
     unsigned char out[8];
     /* Test data from NIST CAVS 14.1 test vectors */
     static unsigned char key[] = {
@@ -1247,7 +1249,7 @@ static int test_kdf_sshkdf(void)
     *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SSHKDF_SESSION_ID,
                                              sessid, sizeof(sessid));
     *p++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_SSHKDF_TYPE,
-                                            &kdftype, sizeof(kdftype));
+                                            kdftype, sizeof(kdftype));
     *p = OSSL_PARAM_construct_end();
 
     ret =
```

File: doc/man7/EVP_KDF-SSHKDF.pod
Comment (t8m) [inline doc/man7/EVP_KDF-SSHKDF.pod:46]: Please fix this typo SSHHKDF when you're at it.

**Comment by t8m** [comment]
Please also fix the `# define EVP_KDF_CTRL_SET_SSHKDF_TYPE        0x11 /* int */` in kdf.h. And maybe also add comments to the individual SSHKDF_TYPE values there like `ASCII A`, and so on.

**Comment by levitte** [comment]
> Please also fix the `# define EVP_KDF_CTRL_SET_SSHKDF_TYPE        0x11 /* int */` in kdf.h. 

Er, that's completely separate, isn't it?  A question, however, is why we have these CTRL macros at all.

**Comment by t8m** [comment]
I got confused by the KDF_CTRL macros. They should not be there at all. And I'd still like to see the comments with the SSHKDF_TYPE values in the header.

**Comment by juergenchrist** [comment]
Shouldn't we also change the macros for the values?  I see the patch from @levitte changes the usage of the macros, but having every user create their own arrays is a little error-prone.  We could do something like
```
#define EVP_KDF_SSHKDF_TYPE_INITIAL_IV_CLI_TO_SRV     (char[]){65, 0}
#define EVP_KDF_SSHKDF_TYPE_INITIAL_IV_SRV_TO_CLI     (char[]){66, 0}
#define EVP_KDF_SSHKDF_TYPE_ENCRYPTION_KEY_CLI_TO_SRV (char[]){67, 0}
#define EVP_KDF_SSHKDF_TYPE_ENCRYPTION_KEY_SRV_TO_CLI (char[]){68, 0}
#define EVP_KDF_SSHKDF_TYPE_INTEGRITY_KEY_CLI_TO_SRV  (char[]) {69, 0}
#define EVP_KDF_SSHKDF_TYPE_INTEGRITY_KEY_SRV_TO_CLI  (char[]) {70, 0}
```
And also add some comments to the values.  Also then, if we adjust the pod-documentation, It really fits together since these values can then directly be used to create the parameter.

Regarding the EVP_KDF_CTRL_SET_SSHKDF_TYPE macro: It seems unused, but I was not sure what it was supposed to be used for.  So I just kept it.  If you want, I can also remove it (and similarly the other EVP_KDF_CTRL_SET_SSHKDF_* macros since they also seem to be unused).

**Comment by t8m** [comment]
> Shouldn't we also change the macros for the values? I see the patch from @levitte changes the usage of the macros, but having every user create their own arrays is a little error-prone. We could do something like
> 
> ```
> #define EVP_KDF_SSHKDF_TYPE_INITIAL_IV_CLI_TO_SRV     (char[]){65, 0}
> #define EVP_KDF_SSHKDF_TYPE_INITIAL_IV_SRV_TO_CLI     (char[]){66, 0}
> #define EVP_KDF_SSHKDF_TYPE_ENCRYPTION_KEY_CLI_TO_SRV (char[]){67, 0}
> #define EVP_KDF_SSHKDF_TYPE_ENCRYPTION_KEY_SRV_TO_CLI (char[]){68, 0}
> #define EVP_KDF_SSHKDF_TYPE_INTEGRITY_KEY_CLI_TO_SRV  (char[]) {69, 0}
> #define EVP_KDF_SSHKDF_TYPE_INTEGRITY_KEY_SRV_TO_CLI  (char[]) {70, 0}
> ```
> 

That makes sense to me.

> And also add some comments to the values. Also then, if we adjust the pod-documentation, It really fits together since these values can then directly be used to create the parameter.
> 
> Regarding the EVP_KDF_CTRL_SET_SSHKDF_TYPE macro: It seems unused, but I was not sure what it was supposed to be used for. So I just kept it. If you want, I can also remove it (and similarly the other EVP_KDF_CTRL_SET_SSHKDF_* macros since they also seem to be unused).

Yep, please do that, possibly as a separate commit as it is an unrelated change.

**Comment by richsalz** [comment]
> #define EVP_KDF_SSHKDF_TYPE_INTEGRITY_KEY_SRV_TO_CLI  (char[]) {70, 0}

Is that portable?

**Comment by levitte** [comment]
Well, the use of macros is counter-intuitive in this case.  If I came from having read the docs, I wonder why I should do more than this:

``` C
char kdftype[] = "A";
...
params++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_SSHKDF_TYPE,
                                            kdftype, sizeof(kdftype));
```

**Comment by levitte** [comment]
> > #define EVP_KDF_SSHKDF_TYPE_INTEGRITY_KEY_SRV_TO_CLI  (char[]) {70, 0}
> 
> Is that portable?

That's definitely not acceptable C90.

**Comment by juergenchrist** [comment]
You are right, but the problem I wanted to avoid is this:
```
char *kdftype = "A";
...
params++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_SSHKDF_TYPE,
                                            kdftype, sizeof(kdftype));
```
I'm very sure this bug will happen...

**Comment by levitte** [comment]
> You are right, but the problem I wanted to avoid is this:

Heh, that is a fairly typical not experienced enough C coder problem.  Mastering that sort of thing is what makes you earn your wings :wink:

**Comment by rsbeckerca** [comment]
AFAIK, the Endian question was addressed in the early 1980s using:

- htons(): host to network short
- htonl(): host to network long
- ntohs(): network to host short
- ntohl(): network to host long

although there was never support for long long, which, I guess is why these fell out of favour. Just saying. That was one of the first things I learned trying to use Berkeley sockets back in (mumblemumble) to talk between machines even when both were little Endian so that network traces would be presentable. Yes, I know it's a non-sequitur, but this is a blocker for my _exotic_ NonStop platform.

File: doc/man7/EVP_KDF-SSHKDF.pod
Comment (juergenchrist) [reply doc/man7/EVP_KDF-SSHKDF.pod]: Done

**Comment by juergenchrist** [comment]
I pushed an updated version with all requested changes.  Please have another look.  @rsbeckerca I can relate.  This is also a blocker for s390x.  This used to be covered via travis.  Not sure if/how this is covered now.

**Comment by t8m** [comment]
Actually this is not an endianness problem per-se. Yes, it demonstrates on big endian machines only but it is actually a problem of misuse of different parameter types here. Maybe that should be reflected in the commit message more prominently - in the first line.

File: include/openssl/kdf.h
Comment (t8m) [inline include/openssl/kdf.h:70]: This is wrong. The "A" etc. should have been just a comment. Otherwise on the EBCDIC machines this will be incorrect.

**Comment by t8m** [comment]
The second commit message is unfortunately also misleading talking about removal of the _TYPE_ values and not about removal of the _CTRL_ values.

**Comment by rsbeckerca** [comment]
> 
> 
> I pushed an updated version with all requested changes. Please have another look. @rsbeckerca I can relate. This is also a blocker for s390x. This used to be covered via travis. Not sure if/how this is covered now.

@juergenchrist On NonStop, it is covered by me building and testing a whole bunch of different configurations on two different architectures and reporting the result back to the team. GitHub does not have a virtual NonStop or s390 box, sadly.

File: include/openssl/kdf.h
Comment (t8m) [reply include/openssl/kdf.h:70]: IMO it would be acceptable to use just {65,0} (without the cast). Then of course you need to adjust the description in the manpage. Something like `Char array initializer of value {65, 0}.`

File: doc/man7/EVP_KDF-SSHKDF.pod
Comment (t8m) [inline doc/man7/EVP_KDF-SSHKDF.pod:106]: `const char type[] = ...`

File: include/openssl/kdf.h
Comment (juergenchrist) [reply include/openssl/kdf.h:70]: As I said above, the alternative is `{ (char)65, (char)0 }`, ...  That should work on all platforms and is a valid C90 array initializer (up to my knowledge).  Any opinions?

Comment (t8m) [reply include/openssl/kdf.h:70]: I would be ok with that.

Comment (juergenchrist) [reply include/openssl/kdf.h]: Done.

**Comment by juergenchrist** [comment]
> The second commit message is unfortunately also misleading talking about removal of the _TYPE_ values and not about removal of the _CTRL_ values.

Sorry, that was a copy/paste error.  Fixed now.

File: doc/man7/EVP_KDF-SSHKDF.pod
Comment (juergenchrist) [reply doc/man7/EVP_KDF-SSHKDF.pod]: Good catch.  Fixed.

**Review by t8m — APPROVED** [review]
LGTM  now!

**Comment by t8m** [comment]
Ping @openssl/committers for the second review.

**Review by levitte — APPROVED** [review]
Yes!

**Review by p-steuer — APPROVED** [review]

**Comment by p-steuer** [comment]
merged.

master: https://github.com/openssl/openssl/commit/d744934b756bc71344818a2cb60b13dd89954afb https://github.com/openssl/openssl/commit/270a5ce1d9ea579a2f1d45887971582b1ef2b6a1

**Pull request closed without merging**
